File: backend/songs/transposer.go
package songs

import (
	"fmt"
	"strings"
)

const (
	C = 0
	D = 2
	E = 4
	F = 5
	G = 7
	A = 9
	B = 11
)

var noteOrder = []int{C, D, E, F, G, A, B}

type Note struct {
	Index      int `json:"index"`      // index of noteOrder
	Alteration int `json:"alteration"` // semitones after or bellow note
}

func (n Note) String() string {
	alteration := ""
	if n.Alteration < 0 {
		alteration = strings.Repeat("♭", n.Alteration*-1)
	} else {
		alteration = strings.Repeat("#", n.Alteration)
	}
	notes := []string{"C", "D", "E", "F", "G", "A", "B"}
	return fmt.Sprintf("%s%s", notes[n.Index], alteration)
}

func (n *Note) AbsoluteCount() int {
	return n.Index + n.Alteration
}

type Tonality struct {
	Name string `json:"name"`
	Note Note   `json:"note"`
}

type Interval struct {
	distanceToNoteIndex int
	semitones           int
}

func (n *Note) transpose(interval Interval) {
	semitones := noteOrder[n.Index] + noteOrder[n.Index+interval.distanceToNoteIndex]
	n.Index = n.Index + interval.distanceToNoteIndex
	n.Alteration = semitones - interval.semitones
}

const transpositionNumber = 12

var sensicalTranspose = []Note{
	{0, 0},  // Dó
	{1, -1}, // Réb
	{1, 0},  // Ré
	{2, -1}, // Mib
	{2, 0},  // Mi
	{3, 0},  // Fá
	{3, 1},  // Fá#
	{4, 0},  // Sol
	{5, -1}, // Láb
	{5, 0},  // Lá
	{6, -1}, // Sib
	{6, 0},  // Si
}

func (t Tonality) TransposeNote(n Note) Interval {
	distanceToNote := t.Note.Index - n.Index
	semitones := n.AbsoluteCount() - t.Note.AbsoluteCount()

	return Interval{distanceToNote, semitones}
}

func (t Tonality) TransposeSemitones(interval int) Interval {
	target := sensicalTranspose[(t.Note.AbsoluteCount()+interval)%transpositionNumber]
	distanceToNote := target.Index - t.Note.Index

	return Interval{distanceToNote, interval}
}

File: frontend/src/SongEditor/song_text/transposer.ts
import { NOTENAMES } from "./songState";

export enum NoteName {
  C = 0,
  D = 2,
  E = 4,
  F = 5,
  G = 7,
  A = 9,
  B = 11,
}

export function noteNameToString(note: NoteName) {
  switch (note) {
    case NoteName.C:
      return "Dó";
    case NoteName.D:
      return "Ré";
    case NoteName.E:
      return "Mi";
    case NoteName.F:
      return "Fá";
    case NoteName.G:
      return "Sol";
    case NoteName.A:
      return "Lá";
    case NoteName.B:
      return "Si";
  }
}

export type Interval = {
  readonly distanceToNoteIndex: number;
  readonly semitoneDistance: number;
};

export type Note = {
  index: number;
  alteration: number;
  absoluteCount: (n: Note) => number;
  transpose: (n: Note, i: Interval) => Note;
  toString: (n: Note) => string;
};

export const NOTESORDER: NoteName[] = [
  NoteName.C,
  NoteName.D,
  NoteName.E,
  NoteName.F,
  NoteName.G,
  NoteName.A,
  NoteName.B,
];

export function newNote(note: NoteName, alteration: number): Note {
  let i = 0;
  while (NOTESORDER[i] !== note) ++i;
  return {
    index: i,
    alteration,
    absoluteCount,
    transpose,
    toString: noteToString,
  };
}

function noteToString(n: Note): string {
  let alteration = "";
  if (n.alteration < 0) {
    alteration += "♭".repeat(n.alteration * -1);
  } else {
    alteration += "#".repeat(n.alteration);
  }
  return NOTENAMES[n.index] + alteration;
}

const absoluteCount = (n: Note) => {
  return NOTESORDER[n.index] + n.alteration;
};

function transpose(n: Note, i: Interval): Note {
  const semitonesAfterIndexing =
    NOTESORDER[i.distanceToNoteIndex] - NOTESORDER[n.index];
  n.index = i.distanceToNoteIndex - n.index;
  n.alteration = semitonesAfterIndexing - NOTESORDER[n.index];
  return n;
}

function transposeNote(from: Note, to: Note): Interval {
  const distanceToNote = from.index - to.index;
  const semitones = absoluteCount(from) - absoluteCount(to);
  return {
    distanceToNoteIndex: distanceToNote,
    semitoneDistance: semitones,
  };
}

export function transposeSemitones(from: Note, st: number): Interval {
  const target = sensicalTranspose[absoluteCount(from) + st];
  return transposeNote(from, target);
}

const sensicalTranspose: Note[] = [
  newNote(NoteName.C, 0), // Dó
  newNote(NoteName.D, -1), // Réb
  newNote(NoteName.D, 0), // Ré
  newNote(NoteName.E, -1), // Mib
  newNote(NoteName.E, 0), // Mi
  newNote(NoteName.F, 0), // Fá
  newNote(NoteName.F, 1), // Fá#
  newNote(NoteName.G, 0), // Sol
  newNote(NoteName.A, -1), // Láb
  newNote(NoteName.A, 0), // Lá
  newNote(NoteName.B, -1), // Sib
  newNote(NoteName.B, 0), // Si
];

export type Tonality = {
  details: string;
  note: Note;
};
